File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.py
def solution(logs):
    answer = []

    # UID별 이름을 담는 사전 -> 현실은 DB로 구현하는게 편할듯?
    nameDict = {}

    # 이름 설정
    for log in logs:
        tmpLog = log.split(sep=' ')

        act = tmpLog[0]
        uid = tmpLog[1]

        if act == "Enter" or act == "Change":
            nameDict[uid] = tmpLog[-1]

    # 로그 파싱 과정
    for log in logs:
        tmpLog = log.split(sep=' ')

        act = tmpLog[0]
        uid = tmpLog[1]

        tmpString = nameDict[uid]

        if act == "Enter":
            tmpString += "님이 들어왔습니다."
        elif act == "Leave":
            tmpString += "님이 나갔습니다."
        else:
            continue

        answer.append(tmpString)

    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <bits/stdc++.h>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    unordered_map<string, string> map;
    
    for(string str : record){
        stringstream ss(str);
        
        string ptr, uid, userName;
        
        ss >> ptr >> uid >> userName;
        
        if(ptr == "Enter" || ptr == "Change") {
            ss >> userName; 
            map[uid] = userName;  
        }    
    }
    
    for(string str : record){
        stringstream ss(str);
        
        string ptr, uid, userName;
        
        ss >> ptr >> uid;
        
        if(ptr == "Enter"){
            string pre = map[uid];
            string In = "님이 들어왔습니다.";
            answer.push_back(pre+In);
        }
        
        else if(ptr == "Leave"){
            string pre = map[uid];
            string Out = "님이 나갔습니다.";
            answer.push_back(pre+Out);
        }
        
    }
    
    return answer;
}